### ISSUE TITLE
Collection size is tied to trial count

### ISSUE THREAD

**Issue by precupqdm** [pholser/junit-quickcheck#187]
We ran into something rather unexpected when switching over from using our own collection generators to the collection generators provided by quickcheck: the collections generated vary depending on how many trials we request. 

As a simplification of the use case, imagine we have an object that we're property testing that contains a List of Lists of objects. Our old list generators had hardcoded sizes, so we never needed to call configure on them. As a result, after the switch, we were using the default sizes from the list generator. We began seeing strange behavior where there were performance issues that were only reproducible on high trial counts, and, after poking into the code, found the problem in PropertyParameterGenerationContext and AbstractGenerationStatus.

It appears the the calculation for the size is as follows, by default:

`return min(distro.sampleWithMean(attempts() + 1, random), sampleSize());`

where `attempts` is the number of trials run so far, and the `sampleSize` is the total number of trials. 

At first, this seemed like a bug, but looking at the source, it seems intentional. What's the intended use of this? As a workaround, we're adding code to call `.configure` on every collection generator, but it seems like this should be another annotation parameter, not something based on the trial count. We didn't expect test runtime to be O(n^2) with respect to trial count. In our case, it's actually much worse, since we have collection of collections and so forth. 

Are we missing something about how to use this or is this intended behavior?

**Comment by csamak** [comment]
I believe this is intended behavior - the notion of increasing "size" is common across most quickcheck implementations (see "The Size of Test Data" in the [QuickCheck manual](http://www.cse.chalmers.se/~rjmh/QuickCheck/manual.html).

As mentioned there, it keeps initial tests fast. Another nice property is that often simple bugs are caught with simple examples even if shrinking isn't implemented (imagine the first test case being an
 n-character string when a 1 character string would have revealed the issue).

**Comment by precupqdm** [comment]
Good point, I see why it makes sense to increase the size as trial runs continue. I still don't understand why the maximum size should be tied to the trial count, though. Running large numbers of trials becomes infeasible because of it.

**Comment by pholser** [comment]
@precupqdm Thanks for your interest! As @csamak surmised, the increasing size is by design. I figured making it roughly track number of attempts was reasonable. I'm open to alternative implementations.

**Comment by pholser** [comment]
@precupqdm As @csamak suggests, another option is to control size of outer and/or inner lists explicitly:

    @Property public void foo(@Size(max = 10) List<@Size(max = 5) List<Integer>> items) {
        // ...
    }

Would this meet your needs? If not, how would you imagine phrasing your properties, ideally?

**Comment by precupqdm** [comment]
What we ended up implementing was a way to specify a default maximum for any collection generated during that particular test. Our setup automatically creates generators for our own types, but they use the generators from this library where possible (including the collection types). That's how we ended up in a situation with multiply nested lists; we generate a list of a type that includes a list of a type that includes a list. This means that we can't directly annotate collections that are within our objects on a per test basis.

I don't think the global per-test maximum was a particularly elegant solution, to be honest, and does stem from our particular setup. It was the best we came up with, though, and has been serving our needs successfully for the past few months.
